### ISSUE TITLE
Implementation Error

### ISSUE THREAD

**Issue by sh-qiangchen** [liwk19/NORA#2]
In equation(9), you use _lp_-norm approximate right part of equation(9).     

![Image](https://github.com/user-attachments/assets/3f1787c3-d638-459c-8ae0-70613d0752b0)

However, the _fr_ is missing in line 193 of main.py.

![Image](https://github.com/user-attachments/assets/e49e0aa1-f48a-42ae-9dfd-58f2ff65c99b)

It may be `new_grad = out * hidden_grad_list[i] * hidden_list[i]`, and it's element-wise multiplication.

**Comment by weikai-li** [comment]
Thank you very much for your suggestion! Actually, we incorporate f_r in Line 184 through the statement "out.backward(gradient=out, retain_graph=True)". Here, the gradient parameter specifies the gradient with respect to the "out" tensor. By passing "out" as the gradient, the inner gradients are multiplied by out.

If we were to perform standard backpropagation without incorporating f_r, we would have used "out.backward(gradient=torch.ones_like(out), retain_graph=True)".

I hope this clarifies your concern. Please let me know if you have any further questions or feedback!

**Comment by sh-qiangchen** [comment]
@weikai-li Thanks! Your work is excellent and easy to follow.

**Comment by weikai-li** [comment]
Thank you very much for your comments!
